core/templates/pages/contributor-dashboard-page/contributions-and-review/contributions-and-review.component.ts

```diff
@@ -344,16 +334,38 @@ angular.module('oppia').component('contributionsAndReview', {
             resolve({opportunitiesDicts: [], more: false});
           });
         }
-        var fetchFunction = tabNameToOpportunityFetchFunction[
-          ctrl.activeSuggestionType][ctrl.activeTabType];
 
-        return fetchFunction().then(function(suggestionIdToSuggestions) {
-          ctrl.contributions = suggestionIdToSuggestions;
-          return {
-            opportunitiesDicts: getContributionSummaries(ctrl.contributions),
-            more: false
-          };
-        });
+        // This function was refactored because after migrating the
+        // contribution-and-review service, its functions were not
+        // getting called properly using the previous method
+        // i.e. fetchFunction().
```

Comment:
Can you please explain the `not getting called properly`?

Reply:
Also, try to avoid changing the structure in the angular migration process.

Reply:
> Can you please explain the `not getting called properly`?

I am not exactly sure why older approach was not working after migration but I am trying to look into it. I first thought that downgrading the `backend-api-service` would make the older approach work but it was not working. Whereas the newer approach is working fine.

Reply:
In the previous approach, we were storing the function of the service (eg. `ContributionAndReviewService.getUserCreatedQuestionSuggestionsAsync`) in the `fetchFunction` and we were calling it. I think that's why the original function of the was not getting called by invoking `fetchFunction` after migration was complete.

I also tried this with `UserService.getUserInfoAsync` which is used in the same file as follows:
```
UserService.getUserInfoAsync().then(function(userInfo) {
  $log.info('getUserInfoAsync');
  ctrl.userIsLoggedIn = userInfo.isLoggedIn();
  ctrl.userDetailsLoading = false;
```
This was working correctly but when we do the following :
```
var fetchFunction = UserService.getUserInfoAsync;

fetchFunction().then(function(userInfo) {
  $log.info('getUserInfoAsync');
  ctrl.userIsLoggedIn = userInfo.isLoggedIn();
```
It fails and we also receive error message in the console.

Reply:
@soumyo123-prog  Can you please share the error log?

Also, can you try with the arrow function? (To me it seems like some scope binding issue (not sure, until I have the proper error log))

Reply:
When we use `fetchFunction` approach with `ContributionAndReviewService` after migration (even after using arrow function syntax), we receive:
```
TypeError: Cannot read properties of undefined (reading '_fetchSuggestionsAsync')
    at eval (contribution-and-review.service.ts:78)
    at Generator.next (<anonymous>)
    at eval (contribution-and-review.service.ts:36)
    at new ZoneAwarePromise (zone.js:917)
    at __awaiter (contribution-and-review.service.ts:32)
    at getUserCreatedTranslationSuggestionsAsync (contribution-and-review.service.ts:77)
    at OpportunitiesListComponent.ctrl.loadContributions [as loadOpportunities] (contributions-and-review.component.ts:295)
    at OpportunitiesListComponent.ngOnInit (opportunities-list.component.ts:80)
    at Object.eval [as next] (opportunities-list.component.ts:65)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:202) TypeError: Cannot read properties of undefined (reading '_fetchSuggestionsAsync')
    at eval (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts?:78:25)
    at Generator.next (<anonymous>)
    at eval (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts?:36:71)
    at new ZoneAwarePromise (webpack:///./node_modules/zone.js/dist/zone.js?:917:33)
    at __awaiter (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts?:32:12)
    at getUserCreatedTranslationSuggestionsAsync (webpack:///./core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts?:77:16)
    at OpportunitiesListComponent.ctrl.loadContributions [as loadOpportunities] (webpack:///./core/templates/pages/contributor-dashboard-page/contributions-and-review/contributions-and-review.component.ts?:295:24)
    at OpportunitiesListComponent.ngOnInit (webpack:///./core/templates/pages/contributor-dashboard-page/opportunities-list/opportunities-list.component.ts?:80:14)
    at Object.eval [as next] (webpack:///./core/templates/pages/contributor-dashboard-page/opportunities-list/opportunities-list.component.ts?:65:67)
    at SafeSubscriber.__tryOrUnsub (webpack:///./node_modules/rxjs/_esm5/internal/Subscriber.js?:202:16)
```

Reply:
And when we use similar approach on `UserService` as in above comment, we get the following error:
```
zone.js:694 Unhandled Promise rejection: Cannot read properties of undefined (reading 'urlService') ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read properties of undefined (reading 'urlService')
    at eval (user.service.ts:70)
    at Generator.next (<anonymous>)
    at eval (user.service.ts:40)
    at new ZoneAwarePromise (zone.js:917)
    at __awaiter (user.service.ts:36)
    at getUserInfoAsync (user.service.ts:69)
    at angular.module.component.controller.ctrl.$onInit (contributions-and-review.component.ts:331)
    at angular.js:10653
    at forEach (angular.js:396)
    at nodeLinkFn (angular.js:10642) TypeError: Cannot read properties of undefined (reading 'urlService')
    at eval (webpack:///./core/templates/services/user.service.ts?:70:35)
    at Generator.next (<anonymous>)
    at eval (webpack:///./core/templates/services/user.service.ts?:40:71)
    at new ZoneAwarePromise (webpack:///./node_modules/zone.js/dist/zone.js?:917:33)
    at __awaiter (webpack:///./core/templates/services/user.service.ts?:36:12)
    at getUserInfoAsync (webpack:///./core/templates/services/user.service.ts?:69:16)
    at angular.module.component.controller.ctrl.$onInit (webpack:///./core/templates/pages/contributor-dashboard-page/contributions-and-review/contributions-and-review.component.ts?:331:17)
    at http://localhost:8181/third_party/static/angularjs-1.8.2/angular.js:10653:34
    at forEach (http://localhost:8181/third_party/static/angularjs-1.8.2/angular.js:396:18)
    at nodeLinkFn (http://localhost:8181/third_party/static/angularjs-1.8.2/angular.js:10642:9)
```

Reply:
Attatching the full error logs:

[log-2.log](https://github.com/oppia/oppia/files/7832799/log-2.log)
[localhost-1641625626449.log](https://github.com/oppia/oppia/files/7832800/localhost-1641625626449.log)

Reply:
Slightly modified `tabNameToOpportunityFetchFunction` variable to use arrow function syntax and it works now